From 2f546bfe5e2dc556ef11c149ab269e39ae073873 Mon Sep 17 00:00:00 2001
From: Andrzej Surdej <Andrzej_Surdej@comcast.com>
Date: Thu, 20 Jun 2024 09:41:20 +0200
Subject: [PATCH 1/2] [DiskCache] Fix WPE_DISK_CACHE_SIZE env parsing

1) Fix WPE_DISK_CACHE_SIZE evn parsing.
   Previously it was falling back to "1" default in all cases.
2) Disable disk cache if no valid number provided
---
 Source/WebKit/Shared/CacheModel.cpp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Source/WebKit/Shared/CacheModel.cpp b/Source/WebKit/Shared/CacheModel.cpp
index df8f69a5e1b2..0f009f523b3e 100644
--- a/Source/WebKit/Shared/CacheModel.cpp
+++ b/Source/WebKit/Shared/CacheModel.cpp
@@ -183,7 +183,7 @@ uint64_t calculateURLCacheDiskCapacity(CacheModel cacheModel, uint64_t diskFreeS
         if (units != 1)
             value = value.substring(0, value.length()-1);
 
-        size_t size = size_t(parseInteger<uint64_t>(s).value_or(1) * units);
+        size_t size = size_t(parseInteger<uint64_t>(value).value_or(0) * units);
         urlCacheDiskCapacity = (unsigned long)(size);
     }
 
-- 
2.25.1


From 19512f72b1149d3ebabafe55121d87af518635d3 Mon Sep 17 00:00:00 2001
From: Andrzej Surdej <Andrzej_Surdej@comcast.com>
Date: Thu, 20 Jun 2024 10:09:04 +0200
Subject: [PATCH 2/2] [DiskCache] Disable HTTP disk cache if not enouhg disk
 space

Hidden behind WPE_DISK_CACHE_SIZE env

There is no control over available disk space for HTTP network cache.
Entries are written to disk asynchronously (using file memory mapping)
so lack of disk space results with SIGBUS signal and WPENetworkProcess crash.
---
 Source/WebKit/Shared/CacheModel.cpp | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/Source/WebKit/Shared/CacheModel.cpp b/Source/WebKit/Shared/CacheModel.cpp
index 0f009f523b3e..010aa8cf5ffb 100644
--- a/Source/WebKit/Shared/CacheModel.cpp
+++ b/Source/WebKit/Shared/CacheModel.cpp
@@ -27,6 +27,7 @@
 #include "CacheModel.h"
 
 #include <algorithm>
+#include <wtf/Assertions.h>
 #include <wtf/RAMSize.h>
 #include <wtf/Seconds.h>
 #include <wtf/StdLibExtras.h>
@@ -183,8 +184,12 @@ uint64_t calculateURLCacheDiskCapacity(CacheModel cacheModel, uint64_t diskFreeS
         if (units != 1)
             value = value.substring(0, value.length()-1);
 
-        size_t size = size_t(parseInteger<uint64_t>(value).value_or(0) * units);
-        urlCacheDiskCapacity = (unsigned long)(size);
+        urlCacheDiskCapacity = parseInteger<uint64_t>(value).value_or(0) * units;
+        if (urlCacheDiskCapacity > diskFreeSize * MB) {
+            WTFLogAlways("Disabling cache due to lack of disk space (wanted space: %ju, disk free space: %ju [bytes])",
+                         urlCacheDiskCapacity, diskFreeSize * MB);
+            urlCacheDiskCapacity = 0;
+        }
     }
 
     return urlCacheDiskCapacity;
-- 
2.25.1

